### PR TITLE
feat(kit): new `maskitoWithPlaceholder` utility

### LINKED ISSUE #181: 🚀 - Add `maskitoWithGuide` processors
### Which package(s) are relevant/related to the feature request?

@maskito/kit

### Description

See this discussion: https://github.com/Tinkoff/maskito/issues/179

We need to add possibility to add guide.

### Proposed solution
```ts
import { Maskito, maskitoPipe, maskitoWithGuide } from '@maskito/core';
import { maskitoWithGuide } from '@maskito/kit';


const { guidePreprocessor, guidePostprocessor } = maskitoWithGuide('dd/mm/yyyy')
 
const maskedInput = new Maskito(element, {
  // ...
  preprocessor: maskitoPipe(
    guidePreprocessor,
    // ...,
    
  ),
  postprocessor: maskitoPipe(
    // ...,
    guidePostprocessor,
  ),
  overwriteMode: 'replace', // ONLY THIS
});
```

**Additional tasks:**
- [ ] Add page `With Guide` to `Recipes` section.
- [ ] Add `Phone` with guide example to this new page
- [ ] Add `Date` with guide example to this new page

### PR DESCRIPTION
**Pull request by nsbarsukov**
## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [X] Feature
- [ ] Refactoring
- [ ] Code style update
- [ ] Build or CI related changes
- [ ] Documentation content changes

## What is the current behavior?

Closes #181

## What is the new behavior?
https://maskito--pr299-guide-qcgp7rbr.web.app/recipes/placeholder

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

### CODE REVIEW

File: projects/demo/src/pages/recipes/placeholder/examples/1-cvc-code/component.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+    readonly maskitoOptions = mask;
+    value = 'xxx';
+
+    constructor(private readonly ngZone: NgZone) {}
+
+    onFocus(): void {
+        const beforePlaceholder = this.value.indexOf('x');
+
+        this.ngZone.runOutsideAngular(() => {
```
Comment (waterplea) [inline projects/demo/src/pages/recipes/placeholder/examples/1-cvc-code/component.ts:43]: Ouch. Do we really need that?

File: projects/kit/src/lib/processors/with-placeholder.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    'postprocessor' | 'preprocessor'
+> & {
+    removePlaceholder: (value: string) => string;
+} {
+    const removePlaceholder = (value: string): string => {
+        for (let i = value.length - 1; i >= 0; i--) {
+            const valueChar = value[i];
+            const placeholderChar = placeholder[i];
+
+            if (valueChar !== placeholderChar) {
```
Comment (waterplea) [inline projects/kit/src/lib/processors/with-placeholder.ts:14]: I think `value[i] !== placeholder[i]` is pretty self-explanatory without constants.

File: projects/core/src/lib/utils/pipe.ts [hunk 3]
```diff
@@ -1,16 +1,19 @@
 import {MaskPostprocessor, MaskPreprocessor} from '../types';
 
-export function maskitoPipe(...processors: readonly MaskPreprocessor[]): MaskPreprocessor;
+export function maskitoPipe(
+    ...processors: ReadonlyArray<MaskPreprocessor | null | undefined>
```
[the remark this answers is not in the document]
Comment (nsbarsukov) [reply projects/core/src/lib/utils/pipe.ts:4]: For DX:

```ts
// I don't know if it contains any proccessors or not
const anyKitMaskOptions = maskitoAnyKitOptionsGenerator({
    // ...
});

export const placeholderOptions = maskitoWithPlaceholder('...');
 
const maskOptions: MaskitoOptions =  {
    ...anyKitMaskOptions,
    preprocessor: maskitoPipe(
        placeholderOptions.preprocessor,
        anyKitMaskOptions.preprocessor, // can be undefined
    ),
    postprocessor: maskitoPipe(
        anyKitMaskOptions.postprocessor, // can be undefined
        placeholderOptions.postprocessor,
    ),
};
```

See last example:
https://maskito--pr299-guide-qcgp7rbr.web.app/recipes/placeholder#date

File: projects/demo/src/pages/recipes/placeholder/examples/1-cvc-code/component.ts (hunk 1 above)
Comment (nsbarsukov) [reply projects/demo/src/pages/recipes/placeholder/examples/1-cvc-code/component.ts]: I've dropped all `runOutsideAngular`

File: projects/kit/src/lib/processors/with-placeholder.ts (hunk 2 above)
Comment (nsbarsukov) [reply projects/kit/src/lib/processors/with-placeholder.ts]: fixed

File: projects/kit/src/lib/processors/with-placeholder.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+            return {
+                elementState: {
+                    selection,
+                    value: removePlaceholder(value),
+                },
+                data,
+            };
+        },
+        postprocessor: ({value, selection}, initialElementState) => {
+            return initialElementState.value
```
[the remark this answers is not in the document]
Comment (nsbarsukov) [reply projects/kit/src/lib/processors/with-placeholder.ts]: > Also unnecessary { return

fixed

**Review by waterplea — APPROVED** [review]
Caret position improvements will be taken care of as https://github.com/Tinkoff/maskito/issues/302

**Pull request merged**